### PR TITLE
validate xml element names

### LINKED ISSUE #7963: XML::Builder allows creating elements with invalid names
Noticed in https://github.com/Blacksmoke16/oq/issues/9.

```cr
require "xml"

xml = XML.build(indent: 2) do |xml|
  xml.element("1") do
    xml.text "foo"
  end
end

puts xml
<?xml version="1.0"?>
<1>foo</1>
```

```cr
require "xml"

xml = XML.build(indent: 2) do |xml|
  xml.element("a b=\"c\"") do
    xml.text "foo"
  end
end

puts xml
<?xml version="1.0"?>
<a b="c">foo</a b="c">
```

I would expect this to error since an element name cannot start with a number nor contain a space, equal sign etc as defined [here](https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-Name).

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Uses `LibXML.xmlValidateNameValue` to validate the provided name when using `XML::Builder` to build elements.

Fixes #7963 
Fixes https://github.com/Blacksmoke16/oq/issues/9

### CODE REVIEW

File: src/xml/builder.cr [hunk 1]
```diff
@@ -44,11 +44,14 @@ struct XML::Builder
...
   def start_element(name : String) : Nil
-    call StartElement, string_to_unsafe(name)
+    element_name = string_to_unsafe name
+    raise XML::Error.new("Invalid element name: '#{name}'", __LINE__) if LibXML.xmlValidateNameValue(element_name).zero?
+    call StartElement, element_name
   end
 
   # Emits the start of an element with namespace info.
   def start_element(prefix : String?, name : String, namespace_uri : String?) : Nil
+    raise XML::Error.new("Invalid element name: '#{prefix}#{name}'", __LINE__) if LibXML.xmlValidateNameValue(string_to_unsafe("#{prefix}#{name}")).zero?
```
Comment (asterite) [inline src/xml/builder.cr]: Shouldn't `prefix` and `name` be validated separately? Or at least, would it be the same to do that? Then we don't need a string interpolation, plus we could say "Invalid prefix" or "Invalid element name".

Comment (Blacksmoke16) [reply src/xml/builder.cr]: Ah, good call.  I'll get that fixed.

File: src/xml/builder.cr [hunk 2]
```diff
@@ -44,14 +44,14 @@ struct XML::Builder
...
-    raise XML::Error.new("Invalid element name: '#{name}'", __LINE__) if LibXML.xmlValidateNameValue(element_name).zero?
-    call StartElement, element_name
+    raise XML::Error.new("Invalid element name: '#{name}'", __LINE__) unless validate_element_name name
+    call StartElement, name
   end
 
   # Emits the start of an element with namespace info.
   def start_element(prefix : String?, name : String, namespace_uri : String?) : Nil
-    raise XML::Error.new("Invalid element name: '#{prefix}#{name}'", __LINE__) if LibXML.xmlValidateNameValue(string_to_unsafe("#{prefix}#{name}")).zero?
+    raise XML::Error.new("Invalid element name: '#{name}'", __LINE__) unless validate_element_name name
```
Comment (asterite) [inline src/xml/builder.cr]: Sorry I forgot to mention this but the `line` argument passed to `XML::Error` is the line number in the XML document. I think that's not available when building so I would pass zero here (just like we pass zero in other places). Probably something to improve later on.

Another idea: have a `check_valid_element_name(name, element_kind)` which does the checking and the raising, so for example like `check_valid_element_name(name, "element name")` and `check_valid_element_name(prefix, "prefix")`.

File: src/xml/builder.cr (hunk 1 above)
Comment (asterite) [reply src/xml/builder.cr]: Thank you! But now `string_to_unsafe` is called twice for each `name` and `prefix`, and that does a null byte checking so now that is done twice. So an optional thing to do is to call that first and then pass those things to `check_valid...`, though I can see it starts to become a bit annoying. Sorry! So feel free to ignore this comment, eventually we can optimize that.

File: src/xml/builder.cr [hunk 3]
```diff
@@ -44,12 +44,19 @@ struct XML::Builder
 
   # Emits the start of an element.
   def start_element(name : String) : Nil
-    call StartElement, string_to_unsafe(name)
+    check_valid_element_name name, string_to_unsafe(name), "element name"
+    call StartElement, name
```
Comment (bcardiff) [inline src/xml/builder.cr]: Isn't this changing the argument?

Shouldn't it be:

```cr
    unsafe_name = string_to_unsafe(name)
    check_valid_element_name name, unsafe_name, "element name"
    call StartElement, unsafe_name
```

Comment (Blacksmoke16) [reply src/xml/builder.cr]: Oh, good catch.  Wonder why that didnt cause things to brek..

Comment (asterite) [reply src/xml/builder.cr]: It's the same

Comment (asterite) [reply src/xml/builder.cr]: Actually it's better the new way to avoid double call

**Pull request merged**